src/policy/v3_policy.cpp

```diff
@@ -0,0 +1,107 @@
+    if (std::all_of(package.cbegin(), package.cend(), [](const auto& tx){return tx->nVersion == 3;})) {
+        return std::nullopt;
+    }
+    // If all transactions are non-V3, we can stop here.
+    if (std::all_of(package.cbegin(), package.cend(), [](const auto& tx){return tx->nVersion != 3;})) {
+        return std::nullopt;
+    }
+    // Look for a V3 transaction spending a non-V3 or vice versa.
+    std::unordered_map<Txid, Wtxid, SaltedTxidHasher> v3_txid_to_wtxid;
+    std::unordered_map<Txid, Wtxid, SaltedTxidHasher> non_v3_txid_to_wtxid;
+    for (const auto& tx : package) {
+        if (tx->nVersion == 3) {
+            v3_txid_to_wtxid.emplace(tx->GetHash(), tx->GetWitnessHash());
+        } else {
+            non_v3_txid_to_wtxid.emplace(tx->GetHash(), tx->GetWitnessHash());
+        }
+    }
+    for (const auto& tx : package) {
+        if (tx->nVersion == 3) {
+            for (const auto& input : tx->vin) {
+                if (auto it = non_v3_txid_to_wtxid.find(input.prevout.hash); it != non_v3_txid_to_wtxid.end()) {
+                    return std::make_tuple(it->second, tx->GetWitnessHash(), true);
+                }
+            }
+        } else {
+            for (const auto& input : tx->vin) {
+                if (auto it = v3_txid_to_wtxid.find(input.prevout.hash); it != v3_txid_to_wtxid.end()) {
+                    return std::make_tuple(it->second, tx->GetWitnessHash(), false);
+                }
+            }
+        }
+    }
+    return std::nullopt;
+}
+
+std::optional<std::string> CheckV3Inheritance(const CTransactionRef& ptx,
+                                              const CTxMemPool::setEntries& ancestors)
+{
+    for (const auto& entry : ancestors) {
+        if (ptx->nVersion != 3 && entry->GetTx().nVersion == 3) {
+            return strprintf("tx that spends from %s must be nVersion=3",
+                             entry->GetTx().GetWitnessHash().ToString());
+        } else if (ptx->nVersion == 3 && entry->GetTx().nVersion != 3) {
+            return strprintf("v3 tx cannot spend from %s which is not nVersion=3",
+                             entry->GetTx().GetWitnessHash().ToString());
+        }
+    }
+    return std::nullopt;
+}
+
+std::optional<std::string> ApplyV3Rules(const CTransactionRef& ptx,
+                                        const CTxMemPool::setEntries& ancestors,
+                                        const std::set<Txid>& direct_conflicts)
+{
+    // These rules only apply to transactions with nVersion=3.
+    if (ptx->nVersion != 3) return std::nullopt;
+
+    if (ancestors.size() + 1 > V3_ANCESTOR_LIMIT) {
+        return strprintf("tx %s would have too many ancestors", ptx->GetWitnessHash().ToString());
+    }
+    if (ancestors.empty()) {
+        return std::nullopt;
+    } else {
+        const auto tx_weight{GetTransactionWeight(*ptx)};
+        // If this transaction spends V3 parents, it cannot be too large.
+        if (tx_weight > V3_CHILD_MAX_WEIGHT) {
```

Comment:
We need to limit the sigops-adjusted size of the child, since that effects the effectiveness of a potential pin. This difference can be up to 20x for really strange scripts.

Simplest but perhaps controversial would be to disallow v3 transactions to have sigops-adjusted size different than bip141 vsize, Other solution would be to directly feed in the adjusted vsize and check against that instead of weight.

Reply:
Yes definitely

> disallow v3 transactions to have sigops-adjusted size different than bip141 vsize

I actually think this is more complicated than adding a sigop limit, since we'd need to define what sigop-adjusted size is. Or just word it as "number of sigops times this constant cannot be larger than the weight" I guess.

Reply:
Added a 50-sigop limit, with a p2sh test.

Reply:
What's the issue with using the transaction's vsize, which is already the value used everywhere else in the mempool?  It seems like that is simpler than introducing a second limit, and exactly captures how we apply the RBF rules anyway.

Reply:
Yes, I think the thing we're actually trying to protect against, sigops-adjusted vsize, is what should be used.

Reply:
Will change 👍